### PR TITLE
[BEAM-13183] [BEAM-8152] Use venv instead of virtualenv to create Python environments in Gradle scripts.

### PR DESCRIPTION
**Pull request by tvalentyn**
Remove a dependency on virtualenv, instead use venv, which is a part of Python library (https://docs.python.org/3/library/venv.html). 

Use venv in gradle commands and in the website documentation.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by tvalentyn** [comment]
Run Python_PVR_Flink PreCommit

**Comment by tvalentyn** [comment]
Run Python_PVR_Flink PreCommit

**Comment by tvalentyn** [comment]
Run PythonDocs PreCommit

**Comment by tvalentyn** [comment]
Run Python_PVR_Flink PreCommit

**Comment by tvalentyn** [comment]
Run Python_PVR_Flink PreCommit

**Comment by tvalentyn** [comment]
Run Python PreCommit

**Comment by tvalentyn** [comment]
Run Python_PVR_Flink PreCommit

**Comment by tvalentyn** [comment]
Run Python PreCommit

**Comment by tvalentyn** [comment]
Run Portable_Python PreCommit

**Comment by tvalentyn** [comment]
Run GoPortable PreCommit

**Comment by tvalentyn** [comment]
Run Python PreCommit

**Comment by tvalentyn** [comment]
Run Portable_Python PreCommit

**Comment by tvalentyn** [comment]
Run Python PreCommit

**Comment by tvalentyn** [comment]
Run Portable_Python PreCommit

**Comment by tvalentyn** [comment]
Run Website PreCommit

**Comment by tvalentyn** [comment]
Run GoPortable PreCommit

**Comment by tvalentyn** [comment]
Run Java PreCommit

**Comment by tvalentyn** [comment]
Run Python_PVR_Flink PreCommit

**Comment by tvalentyn** [comment]
Run SQL PreCommit

**Comment by tvalentyn** [comment]
r: @ryanthompson591 @TheNeuralBit

File: website/www/site/content/en/contribute/_index.md [hunk 1]
```diff
@@ -80,8 +80,9 @@ detail.
    changes locally.
  - For SDK Development:
       - [Go](https://golang.org) 1.12 or later installed for Go SDK development
- - Python 3.6, 3.7, and 3.8. Yes, you need all three versions installed.
-      - pip, setuptools, virtualenv, and tox installed for Python development
+      - Python 3.x interpreters. You will need Python interpreters for all Python versions supported by Beam.
+        Interpreters should be installed and available in shell via `python3.x` commands. For more information, see:
+        Python installation tips in [Developer Wiki](https://cwiki.apache.org/confluence/display/BEAM/Python+Tips#PythonTips-InstallingPythoninterpreters).
```
Comment (TheNeuralBit) [inline website/www/site/content/en/contribute/_index.md:85]: Consider noting the issue with venv in deadsnakes pythons, and the need to install python3.x-dev

File: dev-support/docker/pkglist [hunk 2]
```diff
@@ -33,7 +33,22 @@ openjdk-8-jdk
...
+python3.8-venv
+python3.9
+python3.9-dev
+python3.9-distutils
+python3.9-venv
 tox
 docker.io
+
+
+python3.6 python3.6-dev python3.6-venv python3.7 python3.7-dev python3.7-distutils python3.7-venv python3.8 python3.8-dev  python3.8-distutils python3.8-venv
```
Comment (TheNeuralBit) [inline dev-support/docker/pkglist:54]: what is this line?

File: dev-support/docker/pkglist [hunk 3]
```diff
@@ -33,7 +33,22 @@ openjdk-8-jdk
 python3-setuptools
 python3-pip
 python3.6
+python3.6-dev
+python3.6-venv
```
Comment (TheNeuralBit) [inline dev-support/docker/pkglist:37]: Should we drop 3.6? (also note python is aliased to python3.6 in dev-support/docker/Dockerfile)

File: dev-support/docker/Dockerfile [hunk 4]
```diff
@@ -66,7 +66,7 @@ RUN alias python=python3.6
 ###
 # Install grpcio-tools mypy-protobuf for `python3 sdks/python/setup.py sdist` to work
 ###
-RUN pip3 install grpcio-tools mypy-protobuf virtualenv
+RUN pip3 install grpcio-tools mypy-protobuf
 
 ###
 # Install useful tools
```
Comment (TheNeuralBit) [inline dev-support/docker/Dockerfile:72]: Maybe we can drop distilib below? It says it's installed to workaround a virtualenv issue

File: local-env-setup.sh [hunk 5]
```diff
@@ -85,6 +85,22 @@ elif [ "$kernelname" = "Darwin" ]; then
...
+      if brew ls --versions python@$ver > /dev/null; then
+          echo "python@$ver already installed. Skipping"
+          brew info python@$ver
+      else
+          echo "Installing python@$ver"
+          brew install python@$ver
+      fi
+      if [ ! $(type -P python$ver) > /dev/null 2>&1 ]; then
+        # For some python packages, brew does not add symlinks...
+        # TODO: use pyenv.
```
Comment (TheNeuralBit) [inline local-env-setup.sh:98]: What exactly is the TODO here? Will pyenv make symlinks for us?

File: sdks/python/build-requirements.txt [hunk 6]
```diff
@@ -14,6 +14,12 @@
 #    See the License for the specific language governing permissions and
 #    limitations under the License.
 #
+
+# TODO: Consider PEP-517/PEP-518 instead of this file (BEAM-8954).
```
Comment (TheNeuralBit) [inline sdks/python/build-requirements.txt:18]: nit: 
```suggestion
# TODO(BEAM-8954): Consider PEP-517/PEP-518 instead of this file.
```

File: sdks/python/build-requirements.txt [hunk 7]
```diff
@@ -14,6 +14,12 @@
...
+
+setuptools
+wheel>=0.36.0
```
Comment (TheNeuralBit) [inline sdks/python/build-requirements.txt:21]: why is this change necessary?

File: .test-infra/jenkins/dependency_check/generate_report.sh [hunk 8]
```diff
@@ -49,14 +53,14 @@ rm -f build/dependencyUpdates/beam-dependency-check-report.txt
...
+$PYTHON -m jira_utils.jira_manager_test
+$PYTHON -m dependency_check.version_comparer_test
 
 echo "<html><body>" > $WORKSPACE/src/build/dependencyUpdates/beam-dependency-check-report.html
 
-python -m dependency_check.dependency_check_report_generator Python
+$PYTHON -m dependency_check.dependency_check_report_generator Python
 
-python -m dependency_check.dependency_check_report_generator Java
+$PYTHON -m dependency_check.dependency_check_report_generator Java
```
Comment (TheNeuralBit) [inline .test-infra/jenkins/dependency_check/generate_report.sh:64]: Were you able to verify this (and other .test-infra changes)?

File: release/src/main/scripts/run_rc_validation.sh [hunk 9]
```diff
@@ -648,7 +646,7 @@ if [[ ("$python_xlang_kafka_taxi_dataflow" = true
   do
     rm -rf ./beam_env_${py_version}
     echo "--------------Setting up virtualenv with $py_version interpreter----------------"
-    virtualenv beam_env_${py_version} -p $py_version
+    $py_version -m venv beam_env_${py_version}
```
Comment (TheNeuralBit) [inline release/src/main/scripts/run_rc_validation.sh:649]: Same question about the release script changes, but fortunately you are the next release manager, so I trust you'll be able to detect and fix any issues here.

File: sdks/java/container/build.gradle [hunk 10]
```diff
@@ -22,6 +22,7 @@ plugins {
 }
 
 applyGoNature()
+applyPythonNature()
```
Comment (TheNeuralBit) [inline sdks/java/container/build.gradle:25]: Maybe add a comment that this is needed for `./license_scripts/license_script.sh`. Also, why is it that we need this now, but it seemed to work ok before?

**Review by TheNeuralBit — APPROVED** [review]
This LGTM, I have a few questions though

File: website/www/site/content/en/contribute/_index.md (hunk 1 above)
Comment (tvalentyn) [reply website/www/site/content/en/contribute/_index.md:85]: I added a note to the Wiki about it.

File: dev-support/docker/pkglist (hunk 2 above)
Comment (tvalentyn) [reply dev-support/docker/pkglist:54]: I think there is some 'dev' environment, and this file lists all apt packages to install in that environment.

Comment (tvalentyn) [reply dev-support/docker/pkglist:54]: ah looks like the line is a leftover, thanks for catching.

File: .test-infra/jenkins/dependency_check/generate_report.sh [hunk 11]
```diff
@@ -20,6 +20,9 @@
 set -e
 set -v
 
+# Get currently used Python version from args or assume a default.
+PYTHON=${1:-python3}
```
Comment (lukecwik) [inline .test-infra/jenkins/dependency_check/generate_report.sh:24]: Does this work if users are using an alias/symlink?

File: dev-support/docker/pkglist (hunk 2 above)
Comment (TheNeuralBit) [reply dev-support/docker/pkglist:54]: re: dev envirornment - this line is referenced from `./local-env-setup.sh` (configures current system), and `./start-build-env.sh` (builds and start a development container)

File: .test-infra/jenkins/dependency_check/generate_report.sh (hunk 11 above)
Comment (tvalentyn) [reply .test-infra/jenkins/dependency_check/generate_report.sh:24]: yes, it should work. python3 is typically a symlink.

Comment (tvalentyn) [reply .test-infra/jenkins/dependency_check/generate_report.sh:24]: alias probably wouldn't work

Comment (tvalentyn) [reply .test-infra/jenkins/dependency_check/generate_report.sh:24]: verified that symlinks work and aliases don't

**Comment by tvalentyn** [comment]
Run Seed Job

**Comment by tvalentyn** [comment]
Run Inventory apache-beam-jenkins-1

**Comment by tvalentyn** [comment]
Run Dependency Check

File: dev-support/docker/pkglist (hunk 3 above)
Comment (tvalentyn) [reply dev-support/docker/pkglist:37]: it's to be dropped in Beam 2.36.0.

File: dev-support/docker/Dockerfile (hunk 4 above)
Comment (tvalentyn) [reply dev-support/docker/Dockerfile:72]: I believe I tried to remove it and still ran into some issues. We can try again later.

File: sdks/python/build-requirements.txt (hunk 7 above)
Comment (tvalentyn) [reply sdks/python/build-requirements.txt:21]: these are common build dependencies, however wheel is not installed by default into `venv`, but is installed in virtualenv-created env. without this change, installing from sdist was failing on some python versions.

Comment (tvalentyn) [reply sdks/python/build-requirements.txt:21]: setuptools is still implicitly installed, keeping it explicitly so that we add them when we adopt PEP-517/518 which might require them mentioned explicitly.

File: local-env-setup.sh (hunk 5 above)
Comment (tvalentyn) [reply local-env-setup.sh:98]: Reworded to: # TODO: Consider using pyenv to manage multiple installations of Python.

File: .test-infra/jenkins/dependency_check/generate_report.sh (hunk 8 above)
Comment (tvalentyn) [reply .test-infra/jenkins/dependency_check/generate_report.sh:64]: seems to work, thanks.

File: dev-support/docker/pkglist (hunk 3 above)
Comment (TheNeuralBit) [reply dev-support/docker/pkglist:37]: I see, that makes sense. I thought maybe we were already dropping it because of the `for ver in 3.7 3.8 3.9; do` in `local-env-setup.sh`
